bootloader/grub.go

```diff
@@ -379,11 +386,98 @@ func (g *grub) ManagedAssets() []string {
 	}
 }
 
-// CommandLine returns the kernel command line composed of the built-in
-// list and extra arguments passed in arguments. The command line may be
-// different when using a bootloader in the recovery partition.
+// CommandLine returns the kernel command line composed system run mode
+// arguments, extra arguments (configured by snapd) and the built-in
+// bootloader specific arguments. The command line may be different when
+// using a bootloader in the recovery partition.
 //
 // Implements ManagedAssetsBootloader for the grub bootloader.
-func (g *grub) CommandLine(extra []string) (string, error) {
-	return "", fmt.Errorf("not implemented")
+func (g *grub) CommandLine(modeArgs, extraArgs string) (string, error) {
+	// we do not trust the on disk asset, use the built-in one
+	assetName := "grub.cfg"
+	if g.recovery {
+		assetName = "grub-recovery.cfg"
+	}
+	staticCmdline, err := staticCommandLineFromGrubAsset(assetName)
+	if err != nil {
+		return "", fmt.Errorf("cannot extract static command line element: %v", err)
+	}
+	args, err := strutil.KernelCommandLineSplit(modeArgs + " " + staticCmdline + " " + extraArgs)
```

Comment:
We create the command line with modeArgs first, is that enough to guarantee proper ordering or is there a case where the sort operation below is required? Maybe it could be interesting to also verify if each mode arg is unique to prevent specifying e.g. the operation mode again in the extra args?

Reply:
Perhaps the code should be more strict here. What we want to ensure is that what we assemble here matches the boot config (grub script). However, now that I'm looking at this again, it's clear there's only 2 mode parameters listed in boot config(snapd_recovery_mode & snapd_recovery_system). The code should complain loudly when an unknown parameter is seen in mode args and order the actual one we get appropriately.